{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio import SeqRecord\n",
    "import seaborn as sns\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.cm import viridis\n",
    "import itertools\n",
    "from collections import Counter\n",
    "import re\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Methods section 2.3 and 2.4\n",
    "\n",
    "## derive representative Ty sequences from the CBS432 genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# From assembly, export annotated Ty sequences. To be aligned externally by Muscle\n",
    "#parse genome\n",
    "genome_CBS432 = {}\n",
    "with open('/home/mathieu/mhenault_landrylab/Sequences/ref_genomes/CBS432_pacbio/CBS432.genome.fa') as handle:\n",
    "    for seq in SeqIO.parse(handle, 'fasta'):\n",
    "        genome_CBS432[seq.id] = seq\n",
    "#parse nanotations\n",
    "annot_CBS432 = pd.read_csv('/home/mathieu/mhenault_landrylab/Sequences/ref_genomes/CBS432_pacbio/CBS432.all_feature.gff', sep='\\t', header=None)\n",
    "families_CBS432 = ['TY1','TY3','TY5']\n",
    "fam_seq_CBS432 = {}\n",
    "for TE, df in annot_CBS432.groupby(2):\n",
    "    if TE in families_CBS432:\n",
    "        fam_seq_CBS432[TE] = {}\n",
    "        idx = 0\n",
    "        for i in df.index:\n",
    "            chrom, start, end, strand = df.loc[i, [0,3,4,6]]\n",
    "            ID = f'{idx}-{chrom}-{TE}-{start}-{end}'\n",
    "            seq = genome_CBS432[chrom].seq[start-1:end]\n",
    "            if strand == '-':\n",
    "                seq = seq.reverse_complement()\n",
    "            fam_seq_CBS432[TE][ID] = SeqRecord.SeqRecord(seq=seq, id=ID, description='')\n",
    "            idx += 1\n",
    "        #with open(f'/Users/mathieu/mhenault_landrylab/md_ty_expression/db/{TE}.CBS432.fasta', 'w') as handle:\n",
    "        #    SeqIO.write(fam_seq_CBS432[TE].values(), handle, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import alignments from muscle and generate consensus, to be exported and blasted against annotated sequences\n",
    "alignments_CBS432 = {}\n",
    "for fam, t in itertools.product(families_CBS432, ['internal','LTR']):\n",
    "    with open(f'/home/mathieu/mhenault_landrylab/md_ty_expression/db/{fam}.{t}.CBS432.muscle.fas') as handle:\n",
    "        a = AlignIO.read(handle, 'fasta')\n",
    "    cons = AlignInfo.SummaryInfo(a).dumb_consensus(threshold=0.5, ambiguous='N')\n",
    "    ID = f'cons.{fam}.{t}'\n",
    "    a.add_sequence(ID, str(cons))\n",
    "    alignments_CBS432[ID] = a\n",
    "    #with open(f'/home/mathieu/mhenault_landrylab/md_ty_expression/db/{fam}.{t}.CBS432.cons.fasta', 'w') as handle:\n",
    "    #    AlignIO.write(a, handle, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import blastn results of consensus against annotated sequences\n",
    "# pick winner as representative seq\n",
    "consensus = []\n",
    "for fam, t in itertools.product(families_CBS432, ['internal','LTR']):\n",
    "    ID = f'cons.{fam}.{t}'\n",
    "    blastn = pd.read_csv(f'/home/mathieu/mhenault_landrylab/md_ty_expression/db/{fam}.{t}.CBS432.cons.blastn.tab', header=None, sep='\\t')\n",
    "    \n",
    "    blastn = blastn.loc[(blastn[0]==ID) & (blastn[1]!=ID)].sort_values(by=11, ascending=False)\n",
    "    winner = blastn.iloc[0,1]\n",
    "    # reopen alignment\n",
    "    with open(f'/home/mathieu/mhenault_landrylab/md_ty_expression/db/{fam}.{t}.CBS432.muscle.fas') as handle:\n",
    "        a = AlignIO.read(handle, 'fasta')\n",
    "    a = {seq.id: seq for seq in a}\n",
    "    #add to consensus\n",
    "    \n",
    "    seq = a[winner]\n",
    "    seq.seq = seq.seq.ungap()\n",
    "    print(winner, set(str(seq.seq)))\n",
    "    consensus.append(seq)\n",
    "#with open('/home/mathieu/mhenault_landrylab/md_ty_expression/db/CBS432.consensus.fasta', 'w') as handle:\n",
    "#    SeqIO.write(consensus, handle, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# generate reference sequences with LTR+I for each Spar and Scer genome\n",
    "\n",
    "with open('/home/mathieu/mhenault_landrylab/md_ty_expression/db/CBS432.consensus.fasta', 'r') as handle:\n",
    "    cons_sp = {seq.id:seq for seq in SeqIO.parse(handle, 'fasta')}\n",
    "with open('/home/mathieu/paradoxus_nanopore/paradoxus4/db/tyfamilies_saccharomyces_pone2012.txt', 'r') as handle:\n",
    "    cons_sc = {seq.id.split('#')[0]:seq for seq in SeqIO.parse(handle, 'fasta')}\n",
    "cons = {**cons_sp, **cons_sc}\n",
    "\n",
    "cons_ltr_internal = []\n",
    "for alias, (ltr, internal) in zip(['Ty1_Sp','Ty3_Sp','Ty5_Sp','Ty1_Sc','Ty2_Sc','Ty3_Sc','Ty4_Sc','Ty5_Sc'],\n",
    "                                  [('4-chrVIII-TY1-4262-10147-LTR2', '0-chrIV-TY1-1300771-1306656'),\n",
    "                                   ('1-chrXI-TY3-81086-86445-LTR2', '1-chrXI-TY3-81086-86445'),\n",
    "                                   ('1-chrIII-TY5-9167-14546-LTR1', '1-chrIII-TY5-9167-14546'),\n",
    "                                   ('TY1-LTR', 'TY1-I'),\n",
    "                                   ('TY2-LTR', 'TY2-I'),\n",
    "                                   ('TY3-LTR', 'TY3-I'),\n",
    "                                   ('TY4-LTR', 'TY4-I'),\n",
    "                                   ('TY5-LTR', 'TY5-I')]):\n",
    "    cons_ltr_internal.append(SeqRecord.SeqRecord(seq=(cons[ltr].seq+cons[internal].seq).upper(), id=alias, description=''))\n",
    "\n",
    "with open('/Users/mathieu/mhenault_landrylab/md_ty_expression/db/cons_ltr_internal.fasta', 'w') as handle:\n",
    "    SeqIO.write(cons_ltr_internal, handle, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get length of each LTR\n",
    "for alias, (ltr, internal) in zip(['Ty1_Sp','Ty3_Sp','Ty5_Sp','Ty1_Sc','Ty2_Sc','Ty3_Sc','Ty4_Sc','Ty5_Sc'],\n",
    "                                  [('4-chrVIII-TY1-4262-10147-LTR2', '0-chrIV-TY1-1300771-1306656'),\n",
    "                                   ('1-chrXI-TY3-81086-86445-LTR2', '1-chrXI-TY3-81086-86445'),\n",
    "                                   ('1-chrIII-TY5-9167-14546-LTR1', '1-chrIII-TY5-9167-14546'),\n",
    "                                   ('TY1-LTR', 'TY1-I'),\n",
    "                                   ('TY2-LTR', 'TY2-I'),\n",
    "                                   ('TY3-LTR', 'TY3-I'),\n",
    "                                   ('TY4-LTR', 'TY4-I'),\n",
    "                                   ('TY5-LTR', 'TY5-I')]):\n",
    "    print(alias, len(cons[ltr].seq), len(cons[internal].seq), len(cons[ltr].seq)+len(cons[internal].seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add nuclear genomes to Ty sequences, while masking the former for annotated Tys\n",
    "def parse_mask(S, s, name_append):\n",
    "# parse chromosome and hard-mask\n",
    "    genome_nuc = []\n",
    "    annot = pd.read_csv(f'/Users/mathieu/mhenault_landrylab/Sequences/ref_genomes/{S}_pacbio/{s}.all_feature.gff', sep='\\t', comment='#', header=None)\n",
    "    with open(f'/Users/mathieu/mhenault_landrylab/Sequences/ref_genomes/{S}_pacbio/{s}.genome.fa') as handle:\n",
    "        for seq in SeqIO.parse(handle, 'fasta'):\n",
    "            Seq = seq.seq.tomutable()\n",
    "\n",
    "            # hard-mask anu Ty annotation\n",
    "            for i in annot.loc[(annot[0]==seq.id) & (annot[2].apply(lambda x: 'TY' in x))].index:\n",
    "                start, end = annot.loc[i, [3,4]]\n",
    "                Seq[start-1:end] = 'N'*(end-start+1)\n",
    "            Seq = Seq.toseq()\n",
    "\n",
    "            genome_nuc.append(SeqRecord.SeqRecord(seq=Seq, id=f'{seq.id}_{name_append}', description=''))\n",
    "    return genome_nuc\n",
    "\n",
    "genome_nuc_CBS432 = parse_mask('CBS432','CBS432','Sp')\n",
    "genome_nuc_S288c = parse_mask('S288C','S288c','Sc')\n",
    "\n",
    "# export complete ref\n",
    "with open('/Users/mathieu/mhenault_landrylab/md_ty_expression/db/genome_Sp_Sc_Ty.fasta', 'w') as handle:\n",
    "    SeqIO.write(genome_nuc_CBS432+genome_nuc_S288c+cons_ltr_internal, handle, 'fasta')\n",
    "# export ref with only the Sc Tys\n",
    "with open('/Users/mathieu/mhenault_landrylab/md_ty_expression/db/genome_Sc_Ty.fasta', 'w') as handle:\n",
    "    SeqIO.write(genome_nuc_CBS432+genome_nuc_S288c+cons_ltr_internal[3:], handle, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define parent of genes as their ID to be able to group by Parent\n",
    "def get_parent(x):\n",
    "    patterns = ['(\\w+_\\d{2})[GT](\\d{5})', '(\\w+_Ty\\d)[GT](1)']\n",
    "    value = 'none'\n",
    "    for p in patterns:\n",
    "        m = re.match(p, x)\n",
    "        if m:\n",
    "            value = f'{m.group(1)}.{m.group(2)}'\n",
    "            break\n",
    "    return value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse coordinates of LTRs and CDS for working with depth\n",
    "Coords = pd.read_csv('/home/mathieu/mhenault_landrylab/md_ty_expression/db/ltr_cds_coords.csv', index_col=0)\n",
    "Coords = Coords.drop('Tsu4', axis=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# prepare annotation files for analysis by Plastid"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# for the analysis of periodicity\n",
    "# add pseudo-5' UTR of 50 pb to allow five prime offsets\n",
    "Annot_plastid = []\n",
    "tig_len = {}\n",
    "for S, s, alias in [('CBS432','CBS432','Sp'), ('S288C','S288c','Sc')]:\n",
    "    annot = pd.read_csv(f'/home/mathieu/mhenault_landrylab/Sequences/ref_genomes/{S}_pacbio/{s}.all_feature.gff', sep='\\t', comment='#', header=None)\n",
    "    Annot_plastid.append(annot)\n",
    "    with open(f'/home/mathieu/mhenault_landrylab/Sequences/ref_genomes/{S}_pacbio/{s}.genome.fa') as handle:\n",
    "        tig_len[s] = {tig.id:len(tig.seq) for tig in SeqIO.parse(handle, 'fasta')}\n",
    "# add Ty annotations\n",
    "Annot_plastid.append(pd.read_csv('/home/mathieu/mhenault_landrylab/md_ty_expression/db/annot_ty.gff', sep='\\t', comment='#', header=None))\n",
    "Annot_plastid = pd.concat(Annot_plastid).reset_index(drop=True)\n",
    "Annot_plastid = pd.concat([Annot_plastid, Annot_plastid[8].apply(lambda x: pd.Series(dict([i.split('=') for i in x.split(';')])))], axis=1)\n",
    "Annot_plastid['parent'] = Annot_plastid['ID'].apply(lambda x: get_parent(x))\n",
    "\n",
    "for (p, tig, s, strand), df in Annot_plastid.groupby(['parent',0,1,6]):\n",
    "    \n",
    "    # sort subtables to get the first entries depending on orientation\n",
    "    if strand == '-':\n",
    "        df_sort = df.sort_values(by=4, ascending=False)\n",
    "    if strand == '+':\n",
    "        df_sort = df.sort_values(by=3, ascending=True)\n",
    "        \n",
    "    # iterate and find the first instance of each ['mRNA', 'exon']\n",
    "    df_idx = []\n",
    "    for f in ['mRNA','exon']:\n",
    "        for i in df_sort.index:\n",
    "            if df_sort.loc[i, 2] == f:\n",
    "                df_idx.append(i)\n",
    "                break\n",
    "    \n",
    "    # modify coords accordingly\n",
    "    if strand == '-':\n",
    "        tl = tig_len[s][tig]\n",
    "        Annot_plastid.loc[df_idx, 4] = np.where(df.loc[df_idx, 4]+50<=tl, df.loc[df_idx, 4]+50, tl)\n",
    "    elif strand == '+':\n",
    "        Annot_plastid.loc[df_idx, 3] = np.where(df.loc[df_idx, 3]-50>=1, df.loc[df_idx, 3]-50, 1)\n",
    "    if s == 'CBS432':\n",
    "        Annot_plastid.loc[df.index, 0] = df[0].apply(lambda x: f'{x}_Sp')\n",
    "    if s == 'S288c':\n",
    "        Annot_plastid.loc[df.index, 0] = df[0].apply(lambda x: f'{x}_Sc')\n",
    "\n",
    "#Annot_plastid.loc[Annot_plastid['parent']!='none'].iloc[:, :9].to_csv('/Users/mathieu/mhenault_landrylab/md_ty_expression/plastid/annot_ty_plastid.gff', sep='\\t', header=None, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare gff file for total mRNA quantification\n",
    "# need to consider that Tys will only be scored on internal seq\n",
    "Annot_mrna = []\n",
    "\n",
    "for S, s, alias in [('CBS432','CBS432','Sp'), ('S288C','S288c','Sc')]:\n",
    "    annot = pd.read_csv(f'/home/mathieu/mhenault_landrylab/Sequences/ref_genomes/{S}_pacbio/{s}.all_feature.gff', sep='\\t', comment='#', header=None)\n",
    "    Annot_mrna.append(annot)\n",
    "        \n",
    "Annot_mrna.append(pd.read_csv('/home/mathieu/mhenault_landrylab/md_ty_expression/db/annot_ty.gff', sep='\\t', comment='#', header=None))\n",
    "Annot_mrna = pd.concat(Annot_mrna).reset_index(drop=True)\n",
    "\n",
    "Annot_mrna = pd.concat([Annot_mrna, Annot_mrna[8].apply(lambda x: pd.Series(dict([i.split('=') for i in x.split(';')])))], axis=1)\n",
    "\n",
    "Annot_mrna['parent'] = Annot_mrna['ID'].apply(lambda x: get_parent(x))\n",
    "\n",
    "for s, df in Annot_mrna.groupby(1):\n",
    "    \n",
    "    if s == 'CBS432':\n",
    "        Annot_mrna.loc[df.index, 0] = df[0].apply(lambda x: f'{x}_Sp')\n",
    "    if s == 'S288c':\n",
    "        Annot_mrna.loc[df.index, 0] = df[0].apply(lambda x: f'{x}_Sc')\n",
    "\n",
    "parent_alias = {'CBS432':'Sp', 'S288c':'Sc'}\n",
    "ty_tig_alias = {s: f'{s.split(\"_\")[1][:-2]}_{parent_alias[s.split(\"_\")[0]]}' for s in Coords.index}\n",
    "\n",
    "# extract Ty elms and change start coord to only consider internal sequences\n",
    "for p, df in Annot_mrna.groupby('parent'):\n",
    "    if 'Ty' in p:\n",
    "        start, end = Coords.loc[p.replace('.','T'), ['ltr_end','cds_end']]\n",
    "        Annot_mrna.loc[df.index, [3,4]] = (start+1, end)\n",
    "\n",
    "#Annot_mrna.loc[Annot_mrna['parent']!='none'].iloc[:, :9].to_csv('/Users/mathieu/mhenault_landrylab/md_ty_expression/plastid/annot_ty_plastid_mrna.gff', sep='\\t', header=None, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# prepare gff file for total mRNA quantification, but without Ty1 GAG sequence\n",
    "# same as mrna, but excluding the Ty1 gag part\n",
    "Annot_Ty1 = Annot_mrna.copy()\n",
    "\n",
    "# extract Ty1 elms and change start coord\n",
    "for p, df in Annot_Ty1.groupby('parent'):\n",
    "    if 'Ty1' in p:\n",
    "        start, end = Coords.loc[p.replace('.','T'), ['gag_end','cds_end']]\n",
    "        Annot_Ty1.loc[df.index, [3,4]] = (start+1, end)\n",
    "\n",
    "#Annot_Ty1.loc[Annot_Ty1['parent']!='none'].iloc[:, :9].to_csv('/Users/mathieu/mhenault_landrylab/md_ty_expression/plastid/annot_ty_plastid_Ty1.gff', sep='\\t', header=None, index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add start and end coords for the quantification on the first 60 nts\n",
    "Annot_60nt = Annot_mrna.copy()\n",
    "idx_keep = []\n",
    "for (p, tig, s, strand), df in Annot_60nt.groupby(['parent',0,1,6]):\n",
    "    \n",
    "    start, end = df.iloc[0, [3,4]]\n",
    "\n",
    "    # sort subtables to get the first entries depending on orientation\n",
    "    if strand == '-':\n",
    "        df_sort = df.sort_values(by=4, ascending=False)\n",
    "        start = end - 60\n",
    "    if strand == '+':\n",
    "        df_sort = df.sort_values(by=3, ascending=True)\n",
    "        end = start + 60\n",
    "        \n",
    "    # iterate and find the first instance of each type\n",
    "    df_idx = []\n",
    "    for f in ['gene','mRNA','exon','CDS']:\n",
    "        for i in df_sort.index:\n",
    "            if df_sort.loc[i, 2] == f:\n",
    "                df_idx.append(i)\n",
    "                break\n",
    "    # append indexes to keep for the final clanup\n",
    "    idx_keep.extend(df_idx)\n",
    "    \n",
    "    if strand == '+':\n",
    "        Annot_60nt.loc[df_idx, 4] = end\n",
    "    elif strand == '-':\n",
    "        Annot_60nt.loc[df_idx, 3] = start\n",
    "\n",
    "Annot_60nt = Annot_60nt.loc[idx_keep]\n",
    "\n",
    "#Annot_60nt.loc[Annot_60nt['parent']!='none'].iloc[:, :9].to_csv('/Users/mathieu/mhenault_landrylab/md_ty_expression/plastid/annot_ty_plastid_60nt.gff', sep='\\t', header=None, index=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## analyse the nucleotide identity between Sp and Sc Ty1 and Ty3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# export annotation sequences\n",
    "\n",
    "for TE in ['TY1','TY3']:\n",
    "\n",
    "    seqs = {}\n",
    "\n",
    "    for S, s in zip(['S288C','CBS432'], ['S288c','CBS432']):\n",
    "        genome = {}\n",
    "        with open(f'/home/mathieu/mhenault_landrylab/Sequences/ref_genomes/{S}_pacbio/{s}.genome.fa') as handle:\n",
    "            for seq in SeqIO.parse(handle, 'fasta'):\n",
    "                genome[seq.id] = seq\n",
    "        annot = pd.read_csv(f'/home/mathieu/mhenault_landrylab/Sequences/ref_genomes/{S}_pacbio/{s}.all_feature.gff', sep='\\t', header=None)\n",
    "\n",
    "        for i in annot.loc[annot[2]==TE].index:\n",
    "            chrom, start, end, strand = annot.loc[i, [0,3,4,6]]\n",
    "            ID = f'{s}-{chrom}-{TE}-{start}-{end}'\n",
    "            seq = genome[chrom].seq[start-1:end]\n",
    "            if strand == '-':\n",
    "                seq = seq.reverse_complement()\n",
    "            seqs[ID] = SeqRecord.SeqRecord(seq=seq, id=ID, description='')\n",
    "\n",
    "    #with open(f'/home/mathieu/mhenault_landrylab/md_ty_expression/db/{TE}.fl.fasta', 'w') as handle:\n",
    "    #    SeqIO.write(seqs.values(), handle, 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#define function for identity scoring of one position in a pair \n",
    "#of sequences from a MSA\n",
    "def compare_position(x, gap=np.nan):\n",
    "    if set(x) == '-':\n",
    "        return -1\n",
    "    elif '-' in x:\n",
    "        return -1\n",
    "    elif x[0] != x[1]:\n",
    "        return 0\n",
    "    else:\n",
    "        return 1\n",
    "\n",
    "#define function to summarize identity along a pair of sequences \n",
    "# in a MSA\n",
    "def compare_sequences(sub_a, window, sliding='half'):\n",
    "    cp = np.apply_along_axis(compare_position, 0, sub_a)\n",
    "    cp = np.where(cp!=-1, cp, np.nan)\n",
    "    \n",
    "    if sliding == 'ovl':\n",
    "        return np.array([np.nanmean(cp[i:i+window]) for i in range(cp.shape[0]-window)])\n",
    "    elif sliding == 'half':\n",
    "        half_window = int(window/2)\n",
    "        return np.array([np.nanmean(cp[i:i+window]) for i in range(0, cp.shape[0]-half_window, half_window)])\n",
    "    elif sliding == 'non-ovl':\n",
    "        return np.array([np.nanmean(cp[i:i+window]) for i in range(0, cp.shape[0]-window, window)])\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import alignments and generate consensus\n",
    "\n",
    "SC = {}\n",
    "window = 100\n",
    "for TE in ['TY1','TY3']:\n",
    "\n",
    "    with open(f'/home/mathieu/mhenault_landrylab/md_ty_expression/db/{TE}.fl.muscle.fasta') as handle:\n",
    "        a = AlignIO.read(handle, 'fasta')\n",
    "\n",
    "    sequence_comparisons = {}\n",
    "    idx = 0\n",
    "    with ProgressBar(max_value=360) as bar:\n",
    "        for s_pair in itertools.combinations(a, 2):\n",
    "            species_pair = [x.id.split('-')[0] for x in s_pair]\n",
    "            if len(set(species_pair)) > 1:\n",
    "                sub_a = np.array([x.seq for x in s_pair])\n",
    "                sequence_comparisons['.'.join([x.id for x in s_pair])] = compare_sequences(sub_a, window, sliding='non-ovl')\n",
    "\n",
    "                idx += 1\n",
    "                bar.update(idx)\n",
    "\n",
    "    \n",
    "    df = pd.DataFrame(sequence_comparisons).T\n",
    "    df.columns = (df.columns+0.5)*window\n",
    "    SC[TE] = df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import aln coords of annotations\n",
    "coords_aln = pd.read_csv('/home/mathieu/mhenault_landrylab/md_ty_expression/db/ltr_cds_aln_coords.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plot\n",
    "fig, axes = plt.subplots(nrows=2, figsize=[6,5], \n",
    "                         gridspec_kw={'hspace':1,'left':0.11,'top':0.92,'bottom':0.2,'right':0.97})\n",
    "\n",
    "for TE, alias, alpha, ax in zip(['TY1', 'TY3'], ['Ty1','Ty3'], [0.05,1], axes):\n",
    "    dat = pd.melt(SC[TE])\n",
    "    \n",
    "    ax.plot(dat['variable'], dat['value'], lw=0, marker='o', ms=1, alpha=alpha, c='k', label=TE)\n",
    "    ax.plot(SC[TE].columns, SC[TE].apply(lambda x: np.nanmean(x), axis=0), c='k')\n",
    "    \n",
    "    \n",
    "    ax.axhline(1, ls='--', lw=0.5, c='k')\n",
    "    \n",
    "    ax.set_title(alias)\n",
    "    ax.set_yticks(np.arange(0.6,1.01,0.1))\n",
    "    ax.set_yticklabels(np.arange(60,101,10))\n",
    "    ax.margins(0.01)\n",
    "    ax.set_ylim(0.6, 1.03)\n",
    "    ax.set_ylabel('% nt identity')\n",
    "    \n",
    "    \n",
    "    coords = coords_aln.loc[TE]\n",
    "    \n",
    "    # add patch for five prime LTR\n",
    "    ltr1 = Rectangle([0, 0.32], coords['ltr_end'], 0.05, color='0.3', zorder=0, clip_on=False)\n",
    "    ax.add_patch(ltr1)\n",
    "    ltr2 = Rectangle([coords['ltr_start'], 0.32], coords['end']-coords['ltr_start'], 0.05, color='0.3', zorder=0, clip_on=False)\n",
    "    ax.add_patch(ltr2)\n",
    "    # add patch for internal sequence\n",
    "    internal = Rectangle([coords['ltr_end'], 0.32], coords['ltr_start']-coords['ltr_end'], 0.05, color='0.6', zorder=0, clip_on=False)\n",
    "    ax.add_patch(internal)\n",
    "    # add line for POL\n",
    "    pol = FancyArrow(coords['gag_start'], 0.4, coords['pol_end']-coords['gag_start'], 0, clip_on=False, zorder=1, width=0.04, head_width=0.04, fc='w', ec='k', lw=0.5,\n",
    "                     head_length=50, length_includes_head=True)\n",
    "    ax.add_patch(pol)\n",
    "    ax.text(0.5*(coords['pol_end']+coords['gag_start']), 0.40, '$POL$', va='center', ha='center', color='k', zorder=2, size=6)\n",
    "    # add line for GAG\n",
    "    pol = FancyArrow(coords['gag_start'], 0.45, coords['gag_end']-coords['gag_start'], 0, clip_on=False, zorder=1, width=0.04, head_width=0.04, fc='w', ec='k', lw=0.5,\n",
    "                    head_length=50, length_includes_head=True)\n",
    "    ax.add_patch(pol)\n",
    "    ax.text(0.5*(coords['gag_end']+coords['gag_start']), 0.45, '$GAG$', va='center', ha='center', color='k', zorder=2, size=6)\n",
    "\n",
    "fig.text(0.02, 0.95, 'A', size=18)\n",
    "fig.text(0.02, 0.47, 'B', size=18)\n",
    "    \n",
    "sns.despine()\n",
    "plt.savefig('/home/mathieu/mhenault_landrylab/md_ty_expression/fig/fig_final/FigS2.jpg', dpi=300)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
