{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib import cm\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib.patches import Rectangle\n",
    "from matplotlib.patches import FancyArrow\n",
    "from matplotlib.colors import ListedColormap\n",
    "import seaborn as sns\n",
    "import re\n",
    "from os import listdir\n",
    "import itertools\n",
    "from progressbar import ProgressBar\n",
    "from scipy import stats\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA\n",
    "sns.set(style='ticks', font='Arial')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Methods sections 2.6 and 2.7\n",
    "## import annotations and sample descriptions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import table describing library accessions\n",
    "samples = pd.read_csv('/home/mathieu/mhenault_landrylab/md_ty_expression/hisat2/SraRunTable.txt', header=None)\n",
    "samples = pd.concat([samples, samples[22].apply(lambda x: \\\n",
    "                                                pd.Series(re.match('(.+) ([Ribo|RNA]+-seq) (Rep \\d)', x).groups(),\\\n",
    "                                                          index=['species','experiment','rep']))], axis=1)\n",
    "samples.index = samples[0].values\n",
    "experiment_alias_r = {'RNA-seq':'rna', 'Ribo-seq':'rpf'}\n",
    "species_alias_r = dict(zip(['S. cerevisiae', 'S. paradoxus', 'S. cerevisiae / S. paradoxus F1 Hybrid'],\n",
    "                           ['Scer', 'Spar', 'hyb']))\n",
    "samples['Species'] = samples['species'].apply(lambda x: species_alias_r[x])\n",
    "samples['Experiment'] = samples['experiment'].apply(lambda x: experiment_alias_r[x])\n",
    "samples['Rep'] = samples['rep'].apply(lambda x: f'rep{x[-1]}')\n",
    "\n",
    "samples[[0,'Species','Experiment','Rep']].to_csv('/home/mathieu/mhenault_landrylab/md_ty_expression/deseq2/samples.tab', sep='\\t', index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import gff annotations fed to Plastid\n",
    "Annot_plastid = pd.read_csv('/home/mathieu/mhenault_landrylab/md_ty_expression/plastid/annot_ty_plastid.gff', sep='\\t', header=None)\n",
    "Annot_plastid = pd.concat([Annot_plastid, Annot_plastid[8].apply(lambda x: pd.Series(dict([i.split('=') for i in x.split(';')])))], axis=1)\n",
    "\n",
    "# define parent of genes as their ID to be able to group by Parent\n",
    "def get_parent(x):\n",
    "    patterns = ['(\\w+_\\d{2})[GT](\\d{5})', '(\\w+_Ty\\d)[GT](1)']\n",
    "    value = 'none'\n",
    "    for p in patterns:\n",
    "        m = re.match(p, x)\n",
    "        if m:\n",
    "            value = f'{m.group(1)}.{m.group(2)}'\n",
    "            break\n",
    "    return value\n",
    "\n",
    "Annot_plastid['parent'] = Annot_plastid['ID'].apply(lambda x: get_parent(x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse coordinates of LTRs and CDS per family\n",
    "Coords = pd.read_csv('/home/mathieu/mhenault_landrylab/md_ty_expression/db/ltr_cds_coords.csv', index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define aliases for subgenomes and Ty pseudochromosomes\n",
    "parent_alias = {'CBS432':'Sp', 'S288c':'Sc'}\n",
    "ty_tig_alias = {s: f'{s.split(\"_\")[1][:-2]}_{parent_alias[s.split(\"_\")[0]]}' for s in Coords.index if s != 'Tsu4'}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## import results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# parse plastid count vectors; depth of coverage per position\n",
    "idx = 0\n",
    "Counts = {}\n",
    "with ProgressBar(max_value=11109*12) as bar:\n",
    "    #for s in samples.loc[samples['experiment']=='Ribo-seq'].index:\n",
    "    for s in samples.index:\n",
    "        # define dict to store coverage vectors\n",
    "        counts = {}\n",
    "        # get dir\n",
    "        Dir = f'/home/mathieu/mhenault_landrylab/md_ty_expression/plastid/{s}/'\n",
    "        files = listdir(Dir)\n",
    "        for f in files:\n",
    "            m = re.match('(SRR\\d{6})_((?:S288c|CBS432)_.+[G|T]\\d+)\\.(\\d)', f)\n",
    "            if m:\n",
    "                S, gene, transcript = m.groups()\n",
    "                #with open(, 'r') as handle:\n",
    "                counts[gene] = pd.read_csv(f'{Dir}{f}', header=None, squeeze=True)\n",
    "                idx += 1\n",
    "                bar.update(idx)\n",
    "        Counts[s] = counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import count data from plastid; read counts per gene\n",
    "CT = []\n",
    "for m in ['mrna','60nt','Ty1']:\n",
    "    for s in samples.index:\n",
    "        path = f'/home/mathieu/mhenault_landrylab/md_ty_expression/plastid/{s}/{s}.{m}.counts'\n",
    "        ct = pd.read_csv(path, sep='\\t', header=0, index_col=None, comment='#')\n",
    "        ct['sample'] = s\n",
    "        ct['metric'] = m\n",
    "        CT.append(ct)\n",
    "CT = pd.concat(CT).reset_index(drop=True)\n",
    "CT['gene'] = CT['region_name'].apply(lambda x: x.split('.')[0])\n",
    "for f in ['species','experiment','rep']:\n",
    "    CT[f] = samples.loc[CT['sample'], f].values\n",
    "\n",
    "CT['RPK'] = CT['counts']/CT['length']\n",
    "for s, df in CT.groupby('sample'):\n",
    "    CT.loc[df.index, 'TPM'] = 1e6*df['RPK']/df['RPK'].sum()\n",
    "CT['log_TPM'] = np.log2(CT['TPM']+1)\n",
    "\n",
    "# produce output tables of raw for DEseq\n",
    "#for m, df in CT.groupby('metric'):\n",
    "#    dat = df.pivot_table(index='region_name', columns='sample', values='counts', aggfunc=lambda x: int(x))\n",
    "#    dat.to_csv(f'/home/mathieu/mhenault_landrylab/md_ty_expression/deseq2/{m}.counts.tab',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reformat counts into table with TPM for each replicate\n",
    "TPM = CT.pivot_table(index=['species', 'rep', 'metric', 'region_name'], columns='experiment', values='log_TPM').reset_index()\n",
    "TPM['region_name'] = TPM['region_name'].apply(lambda x: x.split('.')[0])\n",
    "tpm = CT.loc[CT['metric']=='mrna']\n",
    "tpm = pd.Series(tpm['log_TPM'].values, index=pd.MultiIndex.from_frame(tpm[['gene','sample']]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import results from DESeq\n",
    "\n",
    "DESEQ = []\n",
    "path = '/home/mathieu/mhenault_landrylab/md_ty_expression/deseq2/'\n",
    "for f in listdir(path):\n",
    "    m = re.match('results\\.(S288c|CBS432)\\.(\\w+)\\.([a-z_]+)\\.(.+)\\.csv', f)\n",
    "    if m:\n",
    "        subg, metric, lfc, pred = m.groups()\n",
    "        \n",
    "        deseq = pd.read_csv(f'{path}{f}', index_col=0)\n",
    "        deseq['gene'] = deseq.index\n",
    "        deseq['subgenome'] = subg\n",
    "        deseq['metric'] = metric\n",
    "        deseq['lfc'] = lfc\n",
    "        deseq['pred'] = pred\n",
    "        \n",
    "        DESEQ.append(deseq)\n",
    "DESEQ = pd.concat(DESEQ).reset_index(drop=True)\n",
    "DESEQ['log_padj'] = DESEQ['padj'].apply(lambda x: -1*np.log10(x))\n",
    "DESEQ['gene'] = DESEQ['gene'].apply(lambda x: x.split('.')[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import read counts corrected by DESeq normalization\n",
    "counts_norm = []\n",
    "for subg in ['S288c','CBS432']:\n",
    "    data = pd.read_csv(f'/home/mathieu/mhenault_landrylab/md_ty_expression/deseq2/norm_counts.{subg}.mrna.csv', index_col=0)\n",
    "    data = pd.melt(data.reset_index(), id_vars='index', var_name='sample', value_name='count_norm')\n",
    "    data['gene'] = data['index'].apply(lambda x: x.split('.')[0])\n",
    "    data['log_count_norm'] = np.log10(data['count_norm']+1)\n",
    "    data['subgenome'] = subg\n",
    "    counts_norm.append(data)\n",
    "    \n",
    "counts_norm = pd.concat(counts_norm)\n",
    "counts_norm['species'] = samples.loc[counts_norm['sample'], 'species'].values\n",
    "counts_norm['exp'] = samples.loc[counts_norm['sample'], 'experiment'].values\n",
    "counts_norm['Rep'] = samples.loc[counts_norm['sample'], 'Rep'].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import data from DESeq of the RNA-seq only; splitted deseq analysis\n",
    "\n",
    "DESEQ_RNA_SPLIT = []\n",
    "path = '/home/mathieu/mhenault_landrylab/md_ty_expression/deseq2/'\n",
    "for f in listdir(path):\n",
    "    m = re.match('results\\.(\\w+)\\.([a-z_]+)\\.(.+)\\.rnaseq\\.csv', f)\n",
    "    if m:\n",
    "        subgenome, lfc, pred = m.groups()\n",
    "        \n",
    "        deseq = pd.read_csv(f'{path}{f}', index_col=0)\n",
    "        deseq['gene'] = deseq.index\n",
    "        deseq['subgenome'] = subgenome\n",
    "        deseq['lfc'] = lfc\n",
    "        deseq['pred'] = pred \n",
    "        \n",
    "        DESEQ_RNA_SPLIT.append(deseq)\n",
    "DESEQ_RNA_SPLIT = pd.concat(DESEQ_RNA_SPLIT).reset_index(drop=True)\n",
    "DESEQ_RNA_SPLIT['log_padj'] = DESEQ_RNA_SPLIT['padj'].apply(lambda x: -1*np.log10(x))\n",
    "DESEQ_RNA_SPLIT['gene'] = DESEQ_RNA_SPLIT['gene'].apply(lambda x: x.split('.')[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plotting dicts\n",
    "species_encoding = {'Spar':2, 'Scer':0, 'hyb':1}\n",
    "species_order = ['S. cerevisiae', 'S. cerevisiae / S. paradoxus F1 Hybrid', 'S. paradoxus']\n",
    "species_alias = dict(zip(species_order, ['Sc', 'Sc x Sp', 'Sp']))\n",
    "species_color = dict(zip(species_order, ['#0925CF','#07CB18','#F7A220']))\n",
    "\n",
    "Parent_alias = {'CBS432':'par', 'S288c':'cer'}\n",
    "\n",
    "exp_encoding = {'rna':0, 'rpf':1}\n",
    "exp_alias = {'RNA-seq':'mRNA','Ribo-seq':'RPF'}\n",
    "exp_cmap = ListedColormap(['skyblue','steelblue'])\n",
    "exp_markers = {'RNA-seq':'+', 'Ribo-seq':'x'}\n",
    "\n",
    "rep_encoding = {'rep1':0, 'rep2':1}\n",
    "rep_cmap = ListedColormap(['0.2','0.8'])\n",
    "\n",
    "te_order = ['S288c_Ty1T1', 'S288c_Ty2T1', 'S288c_Ty3T1', 'S288c_Ty4T1', 'CBS432_Ty1T1', 'CBS432_Ty3T1', 'CBS432_Ty5T1']\n",
    "Te_order = ['S288c_Ty1T1', 'S288c_Ty2T1', 'S288c_Ty3T1', 'S288c_Ty4T1', 'CBS432_Ty3T1', 'CBS432_Ty5T1']\n",
    "te_alias = {s: f'{s.split(\"_\")[1][:-2]}_{Parent_alias[s.split(\"_\")[0]]}' for s in te_order}\n",
    "te_color = dict(zip(te_order, [cm.tab10(i) for i in range(8)]))\n",
    "te_marker = dict(zip(te_order, ['s','o','^','D','s','^','v']))\n",
    "te_mfc = dict(zip(te_order, ['#07CB18']*4 + [(0,0,0,0)]*3))\n",
    "\n",
    "metric_order = ['60nt', 'mrna', 'Ty1']\n",
    "metric_alias = dict(zip(metric_order, ['60 nt', 'full CDS', 'full CDS/Ty1-$POL$']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot heatmaps of coverage depth z-score per bin, along reference sequences. Fig S1\n",
    "S = samples.copy()\n",
    "S['se'] = S['Species'].apply(lambda x: species_encoding[x])\n",
    "S = S.sort_values(by=['Experiment','se','Rep']).index\n",
    "S_labels = samples.loc[S].apply(lambda x: f'{exp_alias[x[\"experiment\"]]} {species_alias[x[\"species\"]]} {x[\"Rep\"]}', axis=1)\n",
    "wdw = 75\n",
    "\n",
    "# define function to standardize slice of counts\n",
    "def norm_slice(x):\n",
    "    std = x.std()\n",
    "    if std==0:\n",
    "        return pd.Series(np.repeat(np.nan, x.shape[0]), index=x.index)\n",
    "    else:\n",
    "        return (x-x.mean())/x.std()\n",
    "    \n",
    "fig = plt.figure(figsize=[7.8,9])\n",
    "gs = plt.GridSpec(nrows=3, ncols=3, hspace=0.6, wspace=0.8, left=0.12, right=0.96, top=0.96, bottom=0.08)\n",
    "\n",
    "for te, (i,j) in zip(te_order, itertools.product(range(3), range(3))):\n",
    "\n",
    "    Dat = []\n",
    "    for s in S:\n",
    "        dat = pd.DataFrame(Counts[s][te])\n",
    "        dat.columns = ['count']\n",
    "        dat['sample'] = s\n",
    "        dat['pos'] = range(dat.shape[0])\n",
    "        dat['bin'] = pd.cut(dat['pos'], bins=pd.interval_range(start=0, end=dat.shape[0]+wdw, freq=wdw, closed='left'))\n",
    "        Dat.append(dat)\n",
    "    Dat = pd.concat(Dat)\n",
    "    #dat_norm = Dat.groupby(['sample','bin']).mean().reset_index().pivot_table(index='sample', columns='bin', values='count').fillna(0).loc[S]\n",
    "    dat_norm = Dat.groupby(['sample','bin']).mean().reset_index().pivot_table(index='sample', columns='bin', values='count').loc[S]\n",
    "    dat_norm = dat_norm.apply(norm_slice, axis=1)    \n",
    "    #print(te, np.quantile(dat_norm.values.reshape(-1,1), 0.02), np.quantile(dat_norm.values.reshape(-1,1), 0.98))\n",
    "\n",
    "    ax = fig.add_subplot(gs[i,j])\n",
    "  \n",
    "    HM = ax.imshow(dat_norm, cmap='viridis', interpolation='none', aspect='auto', vmin=-1.4, vmax=3.5)\n",
    "    ax.set_yticks(range(12))\n",
    "    ax.set_yticklabels(S_labels, size=6)\n",
    "    \n",
    "    # add TPM\n",
    "    ax.text(1.02, 12.25/12, 'log2\\nTPM', va='bottom', size=6, transform=ax.transAxes)\n",
    "    for pos, s in enumerate(S):\n",
    "        lc = tpm.loc[(te,s)]\n",
    "        ax.text(1.02, (11.5-pos)/12, f'{lc:.2f}', va='center', size=6, transform=ax.transAxes)\n",
    "    \n",
    "    fig.canvas.draw()\n",
    "    tl = [int(t.get_text().replace('âˆ’','-'))*wdw for t in ax.get_xticklabels()]\n",
    "    ax.set_xticklabels(tl)\n",
    "    \n",
    "    # plot TE symbol below heatmap\n",
    "    \n",
    "    base_below = dat_norm.shape[0]-1\n",
    "    unit_below = 0.05*dat_norm.shape[0]\n",
    "    \n",
    "    xrange = ax.axis()[:2]\n",
    "    coords = Coords.loc[te]\n",
    "    slope = (xrange[1]-xrange[0])/(coords['cds_end']-coords.loc['mrna_start'])\n",
    "    intercept = xrange[1]-(coords['cds_end']*slope)\n",
    "    get_data_coord = np.vectorize(lambda x: x*slope+intercept)\n",
    "    coords = get_data_coord(coords[:6])\n",
    "    \n",
    "    # add patch for five prime LTR\n",
    "    ltr = Rectangle([-0.5, base_below+7*unit_below], coords[0], unit_below, color='0.3', zorder=0, clip_on=False)\n",
    "    ax.add_patch(ltr)\n",
    "    # add patch for internal sequence\n",
    "    internal = Rectangle([coords[0], base_below+7*unit_below], coords[1]-coords[0], unit_below, color='0.6', zorder=0, clip_on=False)\n",
    "    ax.add_patch(internal)\n",
    "    # add line for POL\n",
    "    pol = FancyArrow(coords[4], base_below+6*unit_below, coords[1]-coords[4], 0, clip_on=False, zorder=1, width=unit_below, head_width=unit_below, fc='w', ec='k', lw=0.5,\n",
    "                     length_includes_head=True)\n",
    "    ax.add_patch(pol)\n",
    "    ax.text(0.5*(coords[1]-coords[4]), base_below+6*unit_below+0.05, '$POL$', va='center', ha='center', color='k', zorder=2, size=6)\n",
    "    \n",
    "    if 'Ty5' not in te:\n",
    "    # add line for GAG\n",
    "        pol = FancyArrow(coords[4], base_below+4.5*unit_below, coords[2]-coords[4], 0, clip_on=False, zorder=1, width=unit_below, head_width=unit_below, fc='w', ec='k', lw=0.5,\n",
    "                        length_includes_head=True)\n",
    "        ax.add_patch(pol)\n",
    "        ax.text(0.5*(coords[2]-coords[4]), base_below+4.5*unit_below+0.05, '$GAG$', va='center', ha='center', color='k', zorder=2, size=6)\n",
    "    \n",
    "    #for sp in ['left','right', 'top', 'bottom']:\n",
    "    #    ax.spines[sp].set_visible(False)\n",
    "    \n",
    "    ax.set_title(te_alias[te], size=10)\n",
    "    ax.set_xticks(np.arange(0,4501,1500)/75)\n",
    "    ax.set_xticklabels(np.arange(0,4501,1500), size=8)\n",
    "\n",
    "cbar_ax = fig.add_axes([0.55, 0.2, 0.25, 0.02])\n",
    "cbar = plt.colorbar(HM, cax=cbar_ax, orientation='horizontal', label='coverage depth\\nz-score')\n",
    "cbar.outline.set_visible(False)\n",
    "    \n",
    "plt.savefig('/home/mathieu/mhenault_landrylab/md_ty_expression/fig/fig_final/FigS1.jpg', dpi=300)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run PCA on count data\n",
    "sc = StandardScaler()\n",
    "pca = PCA()\n",
    "subg_samples = {'S288c':['S. cerevisiae', 'S. cerevisiae / S. paradoxus F1 Hybrid'],\n",
    "               'CBS432':['S. paradoxus', 'S. cerevisiae / S. paradoxus F1 Hybrid']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# plot PCA results - Fig S3\n",
    "fig = plt.figure(figsize=[7,5.5])\n",
    "gs = plt.GridSpec(ncols=3, nrows=2, wspace=0.7, hspace=0.8, bottom=0.10, left=0.1, top=0.81, right=0.95)\n",
    "\n",
    "for (subg_ax, subg), (metric_ax, m) in itertools.product(enumerate(['S288c', 'CBS432']), enumerate(metric_order)):\n",
    "    \n",
    "    sub_samples = subg_samples[subg]\n",
    "    \n",
    "    dat = CT.loc[(CT['region_name'].apply(lambda x: subg in x))\n",
    "                 & (CT['metric']==m)\n",
    "                 & (CT['species'].isin(sub_samples))]\n",
    "    \n",
    "    dat = dat.pivot_table(index='region_name', columns='sample', values='log_TPM', aggfunc=lambda x: x)\n",
    "    \n",
    "    annot_columns = [f'{species_alias[samples.loc[i,\"species\"]]} \\\n",
    "    {exp_alias[samples.loc[i,\"experiment\"]]} \\\n",
    "    {samples.loc[i,\"rep\"]}' for i in dat.columns]\n",
    "    datn = pd.DataFrame(sc.fit_transform(dat.T.values), index=annot_columns)\n",
    "    \n",
    "    pca_fit = pca.fit(datn)\n",
    "    res = pd.DataFrame(pca_fit.transform(datn), index=dat.columns)\n",
    "    \n",
    "\n",
    "    ax = fig.add_subplot(gs[subg_ax, metric_ax])\n",
    "\n",
    "    # plot paired samples\n",
    "    for (s, rep), df in samples.groupby(['species','rep']):\n",
    "        if np.all(df.index.isin(res.index)):\n",
    "            ax.plot(res.loc[df.index, 0], res.loc[df.index, 1], color=species_color[s])\n",
    "    # plot experiment symbols\n",
    "    for (s, exp), df in samples.groupby(['species','experiment']):\n",
    "        if np.all(df.index.isin(res.index)):\n",
    "            ax.scatter(res.loc[df.index, 0], res.loc[df.index, 1], color=species_color[s], marker=exp_markers[exp], s=48)\n",
    "    \n",
    "    ax.set_title(f'{species_alias[subg_samples[subg][0]]} subgenome\\n{metric_alias[m]}', size=14)\n",
    "    ax.set_xlabel(f'PC1 ({(pca_fit.explained_variance_ratio_[0]*100):.1f} %)')\n",
    "    ax.set_ylabel(f'PC2 ({(pca_fit.explained_variance_ratio_[1]*100):.1f} %)')\n",
    "    ax.margins(0.1)\n",
    "    \n",
    "ax = fig.add_axes([0.2,0.93,0.6,0.04])\n",
    "ax.axis('off')\n",
    "legend_elms = [Line2D([0],[0],color=species_color[s], label=species_alias[s]) for s in species_alias] +\\\n",
    "[Line2D([0],[0], color='w', marker=exp_markers[exp], mec='k', ms=8, label=exp_alias[exp]) for exp in exp_alias]\n",
    "ax.legend(handles=legend_elms, ncol=5, loc=8, frameon=False)\n",
    "    \n",
    "sns.despine()\n",
    "plt.savefig('/home/mathieu/mhenault_landrylab/md_ty_expression/fig/pca_rna_rpf.jpg', dpi=300)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# scatterplot of log TPM for mRNA-seq against RPF - Fig 5A\n",
    "fig = plt.figure(figsize=[7.8,3.5])\n",
    "gs = plt.GridSpec(ncols=3, nrows=2, height_ratios=[2,5], hspace=0.2, wspace=0.4, left=0.1, right=0.95, top=0.97, bottom=0.15)\n",
    "for i, m in enumerate(metric_order):\n",
    "    \n",
    "    df = TPM.loc[TPM['metric']==m]\n",
    "    ax = fig.add_subplot(gs[1,i])\n",
    "    ax.plot(df['RNA-seq'], df['Ribo-seq'], color='0.9', marker='o', ms=5, lw=0, zorder=0)\n",
    "    \n",
    "    for (te, s), df1 in df.loc[df['region_name'].isin(Te_order)].groupby(['region_name','species']):\n",
    "        if 'S288c' in te:\n",
    "            mfc = species_color[s]\n",
    "        if 'CBS432' in te:\n",
    "            mfc = (0,0,0,0)\n",
    "        ax.scatter(df1['RNA-seq'], df1['Ribo-seq'], color=mfc, edgecolors=species_color[s], marker=te_marker[te], s=48, zorder=2)\n",
    "\n",
    "    ax.set_title(metric_alias[m], size=14)\n",
    "    \n",
    "    ax.set_xlabel('log2 TPM mRNA')\n",
    "    if i == 0:\n",
    "        ax.set_ylabel('log2 TPM RPF')\n",
    "    #plot diagonal (reviewer comment)\n",
    "    max_diag = min(ax.axis()[1], ax.axis()[3])\n",
    "    ax.plot([0, max_diag], [0, max_diag], c='k', lw=0.5, zorder=1)\n",
    "    \n",
    "ax = fig.add_subplot(gs[0, :])\n",
    "ax.axis('off')\n",
    "te_fill = dict(zip(te_order, ['k']*4 + [(1,1,1,0)]*3))\n",
    "legend_elms = [Line2D([0], [0], c='w', marker=te_marker[te], mec='k', mfc=te_fill[te], ms=8, label=te_alias[te]) for te in Te_order]\n",
    "L1 = plt.legend(handles=legend_elms, ncol=6, loc=9, bbox_to_anchor=(0.5, 1), frameon=False, columnspacing=0.5)\n",
    "legend_elms = [Line2D([0], [0], c='w', marker='s', mfc=species_color[s], ms=12, label=species_alias[s]) for s in species_order]\n",
    "L2 = plt.legend(handles=legend_elms, ncol=3, loc=8, bbox_to_anchor=(0.5, 0), frameon=False)\n",
    "ax.add_artist(L1)\n",
    "    \n",
    "#fig.text(0.02, 0.78, 'A', size=24, fontweight='bold')\n",
    "    \n",
    "sns.despine()\n",
    "plt.savefig('/home/mathieu/mhenault_landrylab/md_ty_expression/fig/fig_final/Fig5A.jpg', dpi=300)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# stripcharts of deseq-normalized read counts per family - Fig 5B\n",
    "fig = plt.figure(figsize=[7.8,4])\n",
    "gs = plt.GridSpec(ncols=6, nrows=1, wspace=0.7, left=0.07, right=0.85, top=0.88, bottom=0.17)\n",
    "\n",
    "te_ax = dict(zip(Te_order, range(6)))\n",
    "sp_offset = dict(zip(species_order, [0,1,0]))\n",
    "rep_offset = {'rep1':-0.2, 'rep2':0.2}\n",
    "exp_plus = {'RNA-seq':'+', 'Ribo-seq':'x'}\n",
    "\n",
    "for te, df in counts_norm.loc[counts_norm['gene'].isin(Te_order)].groupby('gene'):\n",
    "    ax = fig.add_subplot(gs[te_ax[te]])\n",
    "    \n",
    "    xlabel = {}\n",
    "    for (s, exp, rep), df1 in df.groupby(['species','exp','Rep']):\n",
    "        c = species_color[s]\n",
    "        co = sp_offset[s]\n",
    "        em = exp_plus[exp]\n",
    "        ro = rep_offset[rep]\n",
    "        xlabel[co] = species_alias[s]\n",
    "        \n",
    "        \n",
    "        ax.scatter(co+ro, df1['log_count_norm'].iloc[0], color=c, marker=em, s=48)\n",
    "        \n",
    "    for (s, rep), df1 in df.groupby(['species','Rep']):\n",
    "        c = species_color[s]\n",
    "        co = sp_offset[s]\n",
    "        ro = rep_offset[rep]\n",
    "        \n",
    "        ax.plot(np.repeat(co+ro, 2), df1['log_count_norm'], lw=0.5, c=c)\n",
    "    \n",
    "    if te_ax[te] == 0:\n",
    "        ax.set_ylabel('log2 normalized read counts')\n",
    "\n",
    "    ax.set_xticks(range(2))\n",
    "    ax.set_xticklabels([xlabel[0], xlabel[1]], rotation=90)\n",
    "    ax.margins(0.2)\n",
    "    ax.set_ylim(-0.2, 4.6)\n",
    "    ax.text(0.5, 1.08, te_alias[te], ha='center', size=14, transform=ax.transAxes)\n",
    "    \n",
    "ax = fig.add_axes([0.85,0,0,1])\n",
    "ax.axis('off')\n",
    "legend_elms = [Line2D([0],[0], color='w', marker=exp_markers[exp], mec='k', ms=8, label=exp_alias[exp]) for exp in exp_alias]\n",
    "ax.legend(handles=legend_elms, loc=6, bbox_to_anchor=(1,0.5), frameon=False)\n",
    "\n",
    "#fig.text(0.02, 0.92, 'B', size=24, fontweight='bold')\n",
    "\n",
    "sns.despine()\n",
    "plt.savefig('/home/mathieu/mhenault_landrylab/md_ty_expression/fig/fig_final/Fig5B.jpg', dpi=300)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# volcano plots for the two contrasts of TE efficiency from deseq models - Fig 6\n",
    "\n",
    "fig = plt.figure(figsize=[7.8,7])\n",
    "gs = plt.GridSpec(ncols=3, nrows=3, height_ratios=[0,1,1], hspace=1, wspace=0.5, left=0.1, right=0.97, top=0.98, bottom=0.1)\n",
    "\n",
    "for i, m in enumerate(metric_order):\n",
    "    for (j, subg) in enumerate(['S288c', 'CBS432']):\n",
    "\n",
    "        ax = fig.add_subplot(gs[j+1, i])\n",
    "        df = DESEQ.loc[(DESEQ['metric']==m) &\n",
    "                       (DESEQ['subgenome']==subg) &\n",
    "                       (DESEQ['lfc']=='shrink_lfc') &\n",
    "                       (DESEQ['pred']=='Experimentrpf.Specieshyb')]\n",
    "\n",
    "        df1 = df.loc[~df['gene'].isin(Te_order)]\n",
    "        ax.plot(df1['log2FoldChange'], df1['log_padj'], color='0.9', marker='o', ms=5, lw=0, zorder=0)\n",
    "\n",
    "        for te, df1 in df.loc[df['gene'].isin(Te_order)].groupby('gene'):\n",
    "            ax.scatter(df1['log2FoldChange'], df1['log_padj'], \n",
    "                       marker=te_marker[te], color=te_mfc[te], edgecolor='#07CB18',\n",
    "                       s=48, zorder=1)\n",
    "\n",
    "        for p in (0.05, 0.01):\n",
    "            ax.axhline(-1*np.log10(p), ls='--', lw=1, color='0.5', zorder=0)\n",
    "        ax.axvline(0, ls='--', lw=1, color='0.5', zorder=0)\n",
    "        \n",
    "        ax.set_title(f'{species_alias[subg_samples[subg][0]]} subgenome\\n{metric_alias[m]}', size=14)\n",
    "        ax.set_ylim(-0.2, 3.4)\n",
    "        ax.set_xlim(-2,2)\n",
    "        ax.set_ylabel('-log10 adjusted p-value')\n",
    "        ax.set_xlabel('log2 fold change')\n",
    "        \n",
    "        #add inset\n",
    "        ax_ins = ax.inset_axes((0.7,0.75,0.25,0.25))\n",
    "        ax_ins.plot(df['log2FoldChange'], df['log_padj'], color='k', marker='o', ms=1, lw=0)\n",
    "        # correct xtick labels\n",
    "        ax_ins.set_xlim(-3.5, 3.5)\n",
    "        ax_ins.set_ylim(-0.5, 8)\n",
    "        ax_ins.set_xticks([-3, 3])\n",
    "        ax_ins.set_yticks([0, 5])\n",
    "        ax_ins.set(facecolor=(1,1,1,0))\n",
    "        sns.despine(ax=ax_ins)\n",
    "        \n",
    "        #if i == 2:\n",
    "        if subg == 'S288c':\n",
    "            ax.set_xlabel('log2 fold change\\n(RPF/mRNA : hyb/cer)')\n",
    "        if subg == 'CBS432':\n",
    "            ax.set_xlabel('log2 fold change\\n(RPF/mRNA : hyb/par)')\n",
    "        #ax.text(1.1, 0.5, pred_alias[pred], ha='left', va='center', size=14, rotation=90, transform=ax.transAxes)\n",
    "        ax.margins(0.1)\n",
    "                \n",
    "        \n",
    "ax = fig.add_subplot(gs[0, :])\n",
    "ax.axis('off')\n",
    "legend_elms = [Line2D([0], [0], c='w', marker=te_marker[te], mfc=te_mfc[te], mec='#07CB18', ms=8, label=te_alias[te])\n",
    "               for te in ['S288c_Ty1T1', 'S288c_Ty2T1', 'CBS432_Ty3T1', 'CBS432_Ty5T1']]\n",
    "L1 = plt.legend(handles=legend_elms, ncol=4, loc=9, bbox_to_anchor=(0.5, 1), frameon=False, columnspacing=1)\n",
    "        \n",
    "fig.text(0.02, 0.83, 'A', size=24, fontweight='bold')\n",
    "fig.text(0.02, 0.39, 'B', size=24, fontweight='bold')\n",
    "    \n",
    "sns.despine()\n",
    "plt.savefig('/home/mathieu/mhenault_landrylab/md_ty_expression/fig/fig_final/Fig6.jpg', dpi=300)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# volcano plots for the two contrasts of mRNA expression from deseq - Fig S2\n",
    "\n",
    "fig = plt.figure(figsize=[7,4])\n",
    "gs = plt.GridSpec(ncols=2, nrows=2, height_ratios=[0,1], hspace=0.7, wspace=0.5, left=0.1, right=0.96, top=0.98, bottom=0.18)\n",
    "\n",
    "for i, subg in enumerate(['S288c','CBS432']):\n",
    "\n",
    "    ax = fig.add_subplot(gs[1, i])\n",
    "    df = DESEQ_RNA_SPLIT.loc[(DESEQ_RNA_SPLIT['subgenome']==subg)\n",
    "                             & (DESEQ_RNA_SPLIT['lfc']=='shrink_lfc')\n",
    "                            & (DESEQ_RNA_SPLIT['gene']!='S288c_Ty5T1')]\n",
    "        \n",
    "    df1 = df.loc[~df['gene'].isin(Te_order)]\n",
    "    ax.plot(df1['log2FoldChange'], df1['log_padj'], color='0.9', marker='o', ms=5, lw=0, zorder=0)\n",
    "\n",
    "    for te, df1 in df.loc[df['gene'].isin(Te_order)].groupby('gene'):\n",
    "        ax.scatter(df1['log2FoldChange'], df1['log_padj'], \n",
    "                   marker=te_marker[te], color=te_mfc[te], edgecolor='#07CB18',\n",
    "                   s=48, zorder=1)\n",
    "\n",
    "    for p in (0.05, 0.01):\n",
    "        ax.axhline(-1*np.log10(p), ls='--', lw=1, color='0.5', zorder=0)\n",
    "    ax.axvline(0, ls='--', lw=1, color='0.5', zorder=0)\n",
    "\n",
    "    #ax.set_title(metric_alias[m], size=14)\n",
    "    ax.set_ylim(-0.2, 3.4)\n",
    "    ax.set_xlim(-2,2)\n",
    "    ax.set_ylabel('-log10 adjusted p-value')\n",
    "    ax.set_title(f'{species_alias[subg_samples[subg][0]]} subgenome', size=14)\n",
    "    \n",
    "    if subg == 'S288c':\n",
    "            ax.set_xlabel('log2 fold change\\n(hyb/cer)')\n",
    "    if subg == 'CBS432':\n",
    "        ax.set_xlabel('log2 fold change\\n(hyb/par)')\n",
    "\n",
    "    #add inset\n",
    "    ax_ins = ax.inset_axes((0.7,0.75,0.25,0.25))\n",
    "    ax_ins.plot(df['log2FoldChange'], df['log_padj'], color='k', marker='o', ms=1, lw=0)\n",
    "    # correct xtick labels\n",
    "    ax_ins.set_xlim(-3.5, 3.5)\n",
    "    ax_ins.set_ylim(-2, 40)\n",
    "    ax_ins.set_xticks([-4, 4])\n",
    "    ax_ins.set_yticks([0, 35])\n",
    "    ax_ins.set(facecolor=(1,1,1,0))\n",
    "    sns.despine(ax=ax_ins)\n",
    "\n",
    "ax = fig.add_subplot(gs[0, :])\n",
    "ax.axis('off')\n",
    "legend_elms = [Line2D([0], [0], c='w', marker=te_marker[te], mfc=te_mfc[te], mec='#07CB18', ms=8, label=te_alias[te]) for te in Te_order]\n",
    "L1 = plt.legend(handles=legend_elms, ncol=3, loc=9, bbox_to_anchor=(0.5, 1), frameon=False, columnspacing=1)\n",
    "        \n",
    "sns.despine()\n",
    "plt.savefig('/home/mathieu/mhenault_landrylab/md_ty_expression/fig/fig_final/FigS3.jpg', dpi=300)\n",
    "plt.show()\n",
    "plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
